legate/core/machine.py

```diff
@@ -0,0 +1,279 @@
+                f"{self} & {other}"
+            )
+        return ProcessorRange.create(
+            self.kind,
+            self.per_node_count,
+            max(self.lo, other.lo),
+            min(self.hi, other.hi),
+        )
+
+    def slice(self, sl: slice) -> ProcessorRange:
+        if sl.step is not None and sl.step != 1:
+            raise ValueError("The slicing step must be 1")
+        sz = len(self)
+        new_lo = self.lo
+        new_hi = self.hi
+        if sl.start is not None:
+            if sl.start >= 0:
+                new_lo += sl.start
+            else:
+                new_lo += max(0, sl.start + sz)
+        if sl.stop is not None:
+            if sl.stop >= 0:
+                new_hi = self.lo + sl.stop - 1
+            else:
+                new_hi = self.lo + max(0, sl.stop + sz)
+
+        return ProcessorRange.create(
+            self.kind, self.per_node_count, new_lo, new_hi
+        )
+
+    def __str__(self) -> str:
+        if self.hi < self.lo:
+            return "<empty>"
+        else:
+            return f"[{self.lo}, {self.hi}] ({self.per_node_count} per node)"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def pack(self, buf: BufferBuilder) -> None:
+        buf.pack_32bit_uint(self.per_node_count)
+        buf.pack_32bit_uint(self.lo)
+        buf.pack_32bit_uint(self.hi)
+
+
+ProcKindLike = Union[str, ProcessorKind]
+ProcSlice = Tuple[ProcKindLike, slice]
+
+
+class Machine:
+    def __init__(
+        self,
+        runtime: Runtime,
+        proc_ranges: dict[ProcessorKind, ProcessorRange],
+    ) -> None:
+        self._runtime = runtime
+        self._proc_ranges = proc_ranges
+        self._preferred_kind = ProcessorKind.CPU
+        for kind in ProcessorKind:
+            if kind in proc_ranges and len(proc_ranges[kind]) > 0:
+                self._preferred_kind = kind
+                break
+        self._num_procs = (
+            len(self._proc_ranges[self._preferred_kind])
+            if self._preferred_kind in self._proc_ranges
+            else 0
+        )
+
+    @property
+    def num_procs(self) -> int:
+        return self._num_procs
+
+    @property
+    def preferred_kind(self) -> ProcessorKind:
+        return self._preferred_kind
+
+    def get_processor_range(
+        self, kind: Optional[ProcessorKind] = None
+    ) -> ProcessorRange:
+        if kind is None:
+            sanitized = self._preferred_kind
+        else:
+            sanitized = sanitize_kind(kind)
+        return self._get_range(sanitized)
+
+    def get_node_range(
+        self, kind: Optional[ProcessorKind] = None
+    ) -> tuple[int, int]:
+        proc_range = self.get_processor_range(kind)
+        start_node = proc_range.lo // proc_range.per_node_count
+        end_node = proc_range.hi // proc_range.per_node_count
+        return (start_node, end_node)
+
+    def _get_range(self, kind: ProcessorKind) -> ProcessorRange:
+        if kind not in self._proc_ranges:
```

Comment:
Based on how `only` works, `_get_range` should not error out if a kind is not present in `self._proc_ranges`, it should return an empty `ProcessorRange`.

Reply:
[5898973](https://github.com/nv-legate/legate.core/pull/457/commits/5898973941997efd959a16a5f8cf13118139dd75)